src/avif.c

```diff
@@ -810,7 +797,27 @@ avifBool avifCropRectConvertCleanApertureBox(avifCropRect * cropRect,
     cropRect->y = (uint32_t)(cropY.n / cropY.d);
     cropRect->width = (uint32_t)clapW;
     cropRect->height = (uint32_t)clapH;
-    return avifCropRectIsValid(cropRect, imageW, imageH, yuvFormat, diag);
+    if (!avifCropRectIsValid(cropRect, imageW, imageH, diag)) {
+        return AVIF_FALSE;
+    }
+
+    *upsampleBeforeCropping = ((yuvFormat == AVIF_PIXEL_FORMAT_YUV420 || yuvFormat == AVIF_PIXEL_FORMAT_YUV422) &&
+                               (cropRect->width % 2 || cropRect->x % 2)) ||
+                              (yuvFormat == AVIF_PIXEL_FORMAT_YUV420 && (cropRect->height % 2 || cropRect->y % 2));
```

Comment:
An alternative design is to remove the `yuvFormat` input parameter and the `upsampleBeforeCropping` output parameter from this function (`avifCropRectFromCleanApertureBox`), and then add another function that takes `const avifCropRect * cropRect` and `yuvFormat` as input parameters, and returns `upsampleBeforeCropping` as the return value.

Note that `yuvFormat` and `upsampleBeforeCropping` are only used in this statement (lines 804-806), so this statement can be cleanly moved to a separate function.

Another alternative is for libavif to upsample to 4:4:4 and crop automatically when necessary so that libavif users don't need to worry about this. But I am not sure if this is a good idea.

Reply:
> add another function that takes `const avifCropRect * cropRect` and `yuvFormat` as input parameters, and returns `upsampleBeforeCropping` as the return value

The issue with this new function and with the current `avifImage::clap` are that users can easily ignore them. The point of the "mandatory" `upsampleBeforeCropping` argument was to force users to at least consider upsampling (because there is no other function available to get a rect). But they can just ignore `clap` anyway so it has little effect.

> Another alternative is for libavif to upsample to 4:4:4 and crop automatically when necessary so that libavif users don't need to worry about this. But I am not sure if this is a good idea.

I would lean towards that as a default behavior, and why not add an "ignoreClap" setting in `avifDecoder` to keep the old behavior back.  
However existing files with `clap` will no longer display the same, in places where `clap` is ignored by users of libavif.

Reply:
> An alternative design is to remove the `yuvFormat` input parameter and the `upsampleBeforeCropping` output parameter from this function (`avifCropRectFromCleanApertureBox`), and then add another function that takes `const avifCropRect * cropRect` and `yuvFormat` as input parameters, and returns `upsampleBeforeCropping` as the return value.
> 
> Note that `yuvFormat` and `upsampleBeforeCropping` are only used in this statement (lines 804-806), so this statement can be cleanly moved to a separate function.

Done.